Gems/ROS2/Code/Source/Manipulator/ManipulatorControllerComponent.cpp

```diff
@@ -0,0 +1,330 @@
+        m_actionServerClass.m_actionServer.reset();
+    }
+
+
+    void ManipulatorControllerComponent::GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required)
+    {
+        required.push_back(AZ_CRC_CE("JointPublisher"));
+    }
+
+
+    void ManipulatorControllerComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serialize->Class<ManipulatorControllerComponent, AZ::Component>()
+                ->Version(0)
+                ->Field("Controller type", &ManipulatorControllerComponent::m_controllerType)
+                ->Field("PID Configuration Vector", &ManipulatorControllerComponent::m_pidConfigurationVector);
+
+            if (AZ::EditContext* ec = serialize->GetEditContext())
+            {
+                ec->Class<ManipulatorControllerComponent>("ManipulatorControllerComponent", "[Controller for a robotic arm (only hinge joints)]")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC("Game"))
+                    ->Attribute(AZ::Edit::Attributes::Category, "ROS2")
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::ComboBox,
+                        &ManipulatorControllerComponent::m_controllerType,
+                        "Controller type", 
+                        "Different controller types to command the joints of the manipulator")
+                    ->EnumAttribute(ManipulatorControllerComponent::Controller::FeedForward, "FeedForward")
+                    ->EnumAttribute(ManipulatorControllerComponent::Controller::PID, "PID")
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default, 
+                        &ManipulatorControllerComponent::m_pidConfigurationVector,
+                        "PIDs Configuration", 
+                        "PID controllers configuration (for all the joints)");
+            }
+        }
+    }
+
+    void ManipulatorControllerComponent::InitializeMap()
+    {
+        AZStd::vector<AZ::EntityId> descendants;
+        AZ::TransformBus::EventResult(descendants, GetEntityId(), &AZ::TransformInterface::GetAllDescendants);
+
+        for (const AZ::EntityId& descendantID : descendants)
+        {
+            AZ::Entity* entity = nullptr;
+            AZ::ComponentApplicationBus::BroadcastResult(entity, &AZ::ComponentApplicationRequests::FindEntity, descendantID);
+            AZ_Assert(entity, "Unknown entity %s", descendantID.ToString().c_str());
+            auto* frameComponent = entity->FindComponent<ROS2FrameComponent>();
+            auto* hingeComponent = entity->FindComponent<PhysX::HingeJointComponent>();
+            if (frameComponent && hingeComponent)
+            {
+                m_hierarchyMap[frameComponent->GetJointName()] = descendantID;
```

Comment:
I am still building a map to store the entityID and the jointName at the launching of the Game mode. Any comments on how to avoid that or better/more efficient solutions?